src/librustc_middle/mir/mod.rs

```diff
@@ -2935,4 +2951,14 @@ pub struct CoverageData {
 
     /// The total number of coverage region counters added to the MIR `Body`.
     pub num_counters: u32,
+
+    /// The start and end positions within a source file for the region of source code counted by
+    /// the given counter index.
+    pub coverage_regions: IndexVec<u32, CoverageRegion>,
```

Comment:
I saw your comment about gap regions below.. sounds like clang tackling the same (or a similar) problem.

Reply:
Sounds like you believe the gap regions may address your "one-to-many" question. Is that right?

If you're not sure, can you clarify what you mean by one-to-many? I didn't follow it. (For instance, I'm not sure what type/thing is the "one" and what are the "many"?)

But if it helps, the coverage map includes:

1. A list of counter indexes for each injected counter call (instrprof.increment), with the span of code (region) that executed for each increment of this counter.
2. A second list of counter expressions with their code regions, for each region that can be counted by extrapolation from other counter and/or counter expression values.
3. Zero-counters (representing unreachable statements, for example)

Reply:
What I mean is that a counter could somehow be associated with non-contiguous sections of code in the presence of optimizations. Say something like

```rust
if (foo) {
  do_stuff();
}

let x = 2;

if (foo) {
  do_other_stuff();
}

println!("{}", x);
```

The compiler could easily reorder and collapse the if bodies into a single branch. For now I don't think we'll encounter this (we should create counters before any such optimization, and if the optimization happens then we'll just increment multiple counters in the combined branch).

I'm only posing the question of whether this is something that we might want to handle, or if there are any cases where this could happen today. I can't think of any right now.

Reply:
Ah, yes, that makes sense. I expect the first iteration of coverage analysis will inject counters in each `if` block, but if there's a way to identify the scenario above automatically, the second one could be replaced with a counter expression (`first_counter` + 0) at no runtime cost (with or without the optimizer merging the two). That seems hard though, but I also wonder if this happens enough to make it worth trying to optimize this specific case.

Something to keep in mind though.

I'll resolve this conversation for now. Thanks!